wip/services/exec/store.go

```diff
@@ -18,55 +17,128 @@ func NewPostgresStore() Store {
 	if err != nil {
 		return fmt.Errorf("failed to marshal spec to JSON: %w", err)
 	}
+
 	metadata, err := json.Marshal(&types.NodeMetadataV1{})
 	if err != nil {
 		return fmt.Errorf("failed to marshal metadata to JSON: %w", err)
 	}
 
-	dbp := db.ExecCreateParams{
-		ID:           exec.ID,
-		CreatedBy:    exec.CreatedBy,
-		ProjectID:    project,
-		Region:       exec.Region,
-		Name:         exec.Name,
-		Spec:         spec,
-		Metadata:     metadata,
-		CommitID:     commitID,
-		GitRemoteUrl: gitRemoteURL,
-		Command:      command,
-		BuildID:      bid,
-		Image:        exec.Image,
-		Provider:     exec.Provider.String(),
-	}
-
-	if err := db.Q.ExecCreate(context.Background(), dbp); err != nil {
-		// TODO: parse db errors: project not found, ssh key not found,
+	err = createSSHKeys(ctx, exec.CreatedBy, publicKeys)
+	if err != nil {
+		return err
+	}
+	err = createExec(project, spec, metadata, exec)
+	if err != nil {
+		return err
+	}
+	keys, err := getSSHKeysByPublicKey(ctx, exec.CreatedBy, types.GetPublicKeys(publicKeys))
+	if err != nil {
+		return err
+	}
+	err = addSSHKeyToExec(ctx, exec, keys)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createExec(projectID string, spec []byte, metadata []byte, exec types.Exec) error {
+	if err := db.Q.ExecCreate(context.Background(), db.ExecCreateParams{
+		ID:        exec.ID,
+		CreatedBy: exec.CreatedBy,
+		ProjectID: projectID,
+		Region:    exec.Region,
+		Name:      exec.Name,
+		Spec:      spec,
+		Metadata:  metadata,
+		BuildID:   db.NullStringFrom(exec.BuildID),
+		Image:     exec.Image,
+		Provider:  exec.Provider.String(),
+
+		// Note: These fields are members of the Exec, but currently unused in any feature.
+		CommitID:     db.NullStringFrom(exec.CommitID),
+		GitRemoteUrl: db.NullStringFrom(exec.GitURL),
+		Command:      []string{},
+	}); err != nil {
 		return err
 	}
+
+	return nil
+}
+
+func createSSHKeys(ctx context.Context, createdByID string, keys []types.SSHKey) error {
+	for _, key := range keys {
+		exists, err := db.Q.SSHKeyGetByPublicKey(ctx, db.SSHKeyGetByPublicKeyParams{
+			PublicKey: *key.PublicKey,
+			OwnerID:   createdByID,
+		})
+		if err != nil && err != sql.ErrNoRows {
+			return err
+		}
+		if exists.PublicKey != "" {
+			continue
+		}
+		err = db.Q.SSHKeyAdd(ctx, db.SSHKeyAddParams{
+			OwnerID:   createdByID,
+			Name:      key.Name,
+			PublicKey: *key.PublicKey,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getSSHKeysByPublicKey(ctx context.Context, ownerID string, pubs []string) ([]db.UnweaveSshKey, error) {
+	keys := make([]db.UnweaveSshKey, 0, len(pubs))
+
+	for _, pub := range pubs {
+		key, err := db.Q.SSHKeyGetByPublicKey(ctx, db.SSHKeyGetByPublicKeyParams{
+			PublicKey: pub,
+			OwnerID:   ownerID,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		keys = append(keys, key)
+	}
+
+	return keys, nil
+}
+
+func addSSHKeyToExec(ctx context.Context, exec types.Exec, keys []db.UnweaveSshKey) error {
```

Comment:
Let's not mix methods and private functions in the file. Putting them at the bottom of the file is the convention I've been using

Reply:
Np +1. Normally I’d have the related private methods closer to associated structures or functionality